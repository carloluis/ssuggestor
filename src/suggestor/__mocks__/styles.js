const styles = require.requireActual('../styles');

export const SPIN_STYLES = styles.SPIN_STYLES;

export const X_STYLES = styles.X_STYLES;

export const getListStyles = jest.fn(styles.getListStyles);
