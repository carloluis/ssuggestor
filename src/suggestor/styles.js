export const SPIN_STYLES = {
	position: 'absolute',
	cursor: 'pointer',
	margin: 'auto',
	color: '#ccc',
	fontSize:16,
	right:10,
	top: 0,
	bottom: 0,
	height: 14,
	zIndex:4
};

export const X_STYLES = {
	...SPIN_STYLES,
	fontSize: 14,
	right: 30
};

export const getListStyles = visible => ({
	display: visible? 'block':'none',
	maxHeight: 250,
	minWidth: 30,
	width: 'inherit',
	overflow: 'auto',
	cursor: 'pointer'
});