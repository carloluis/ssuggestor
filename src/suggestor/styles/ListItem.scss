:global .ss-over-item {
	background: #eee;
}
