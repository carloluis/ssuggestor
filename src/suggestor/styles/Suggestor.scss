:global .ss-triangle,
:global .ss-remove {
	position: absolute;
	cursor: pointer;
	margin: auto;
	color: #ccc;
	font-size: 16px;
	right: 10px;
	top: 0;
	bottom: 0;
	height: 14px;
	z-index: 4;
}

:global .ss-remove {
	font-size: 14px;
	right: 30px;
}
