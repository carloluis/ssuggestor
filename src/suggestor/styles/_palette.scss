$pale-grey: #eee;
$grey: #ccc;
