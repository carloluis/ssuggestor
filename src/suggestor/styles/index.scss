@import './Suggestor.scss';
@import './List.scss';
