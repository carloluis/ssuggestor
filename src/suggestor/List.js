import React from 'react';
import ListItem from './ListItem';
import { getListStyles } from './styles';

export const List = ({ open, filtered, index, value, onItemClick, onItemMouseEnter }) => (
	(open && !!filtered.length) && 
	<ul className="dropdown-menu" style={getListStyles(open)}>
		{ filtered.map((item, i) => 
			<ListItem key={item.word} {...{ item, onItemClick, onItemMouseEnter, search:value, index:i, overItem:i===index }} />) }
	</ul>
);

List.propTypes = {
	filtered: React.PropTypes.array.isRequired,
	index: React.PropTypes.number.isRequired,
	value: React.PropTypes.string.isRequired,
	open: React.PropTypes.bool.isRequired,
	onItemClick: React.PropTypes.func.isRequired,
	onItemMouseEnter: React.PropTypes.func.isRequired
};

export default List;