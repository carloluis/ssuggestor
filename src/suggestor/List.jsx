import React from 'react';
import PropTypes from 'prop-types';
import ListItem from './ListItem';

const List = ({ open, filtered, index, value, onItemClick, onItemMouseEnter, classSchema }) =>
	open &&
	!!filtered.length && (
		<ul className={classSchema.list}>
			{filtered.map((item, i) => (
				<ListItem
					key={item.word}
					{...{
						item,
						onItemClick,
						onItemMouseEnter,
						search: value,
						index: i,
						overItem: i === index,
						classSchema
					}}
				/>
			))}
		</ul>
	);

List.propTypes = {
	filtered: PropTypes.array.isRequired,
	index: PropTypes.number.isRequired,
	onItemClick: PropTypes.func.isRequired,
	onItemMouseEnter: PropTypes.func.isRequired,
	value: PropTypes.string.isRequired,
	open: PropTypes.bool.isRequired,
	classSchema: PropTypes.shape({
		list: PropTypes.string,
		item: PropTypes.string,
		activeItem: PropTypes.string
	})
};

export default List;
