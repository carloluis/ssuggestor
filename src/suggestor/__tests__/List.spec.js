import React from 'react';
import List from '../List';
import { getListStyles } from '../styles';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

jest.mock('../styles');

const PROPS = {
	filtered: [],
	index: 0,
	onItemClick: jest.fn(),
	onItemMouseEnter: jest.fn(),
	open: false,
	value: ''
};

describe('<List />', () => {
	const filtered = [{ word: 'locellus', index: 0 }, { word: 'Lydian', index: 0 }, { word: 'warray', index: -1 }];
	const propsWithItems = { ...PROPS, filtered };

	it('snapshot', () => {
		const tree = renderer.create(<List {...PROPS} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - no suggestion list', () => {
		const tree = renderer.create(<List {...PROPS} open />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('snapshot - suggestion list closed', () => {
		const tree = renderer.create(<List {...propsWithItems} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('snapshot - suggestion list open', () => {
		const tree = renderer.create(<List {...propsWithItems} open />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('snapshot - suggestion list open (index 1)', () => {
		const props = { ...PROPS, filtered, index: 1 };

		const tree = renderer.create(<List {...props} open />).toJSON();

		expect(tree).toMatchSnapshot();
	});
});

describe('List component', () => {
	const filtered = [{ word: 'locellus', index: 0 }, { word: 'Lydian', index: 0 }, { word: 'warray', index: -1 }];

	beforeEach(() => {
		getListStyles.mockReset();
		PROPS.onItemClick.mockClear();
		PROPS.onItemMouseEnter.mockClear();
	});

	it('should render nothing when no filtered items', () => {
		const wrapper = shallow(<List {...PROPS} />);

		expect(wrapper.getElement()).toBeFalsy();
		expect(wrapper.equals(false)).toBeTruthy();
		expect(wrapper.find('.dropdown-menu').getElements().length).toBeFalsy();
	});

	it('should render nothing when filtered items are hidden', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} />);

		expect(wrapper.getElement()).toBeFalsy();
	});

	it('should render when open flag and any filtered items', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} open />);

		expect(wrapper.getElement()).toBeTruthy();
		expect(wrapper.find('.dropdown-menu').getElement()).toBeTruthy();
	});

	it('should render each item inside an <li> element', () => {
		const wrapper = mount(<List {...PROPS} filtered={filtered} open />);

		const lis = wrapper.find('li');

		expect(lis.length).toBe(3);
	});

	it('should render each item inside an <li> element', () => {
		const wrapper = mount(<List {...PROPS} filtered={filtered} open />);

		const lis = wrapper.find('li');

		expect(lis.first().text()).toBe(filtered[0].word);
	});

	it('should call onItemClick prop when <ListItem> is clicked', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} open />);

		wrapper.find('ListItem').first().simulate('itemClick');

		expect(PROPS.onItemClick).toHaveBeenCalledTimes(1);
	});

	it('should call onItemClick prop when each <ListItem> is clicked', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} open />);

		wrapper.find('ListItem').first().simulate('itemClick');
		wrapper.find('ListItem').at(1).simulate('itemClick');
		wrapper.find('ListItem').last(2).simulate('itemClick');

		expect(PROPS.onItemClick).toHaveBeenCalledTimes(3);
	});

	it('should call onItemMouseEnter prop when mouse enter into <ListItem> area', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} open />);

		wrapper.find('ListItem').first().simulate('itemMouseEnter');

		expect(PROPS.onItemMouseEnter).toHaveBeenCalledTimes(1);
	});

	it('should call onItemMouseEnter prop when mouse enter into each <ListItem> area', () => {
		const wrapper = shallow(<List {...PROPS} filtered={filtered} open />);

		wrapper.find('ListItem').first().simulate('itemMouseEnter');
		wrapper.find('ListItem').at(1).simulate('itemMouseEnter');
		wrapper.find('ListItem').last().simulate('itemMouseEnter');

		expect(PROPS.onItemMouseEnter).toHaveBeenCalledTimes(3);
	});
});
