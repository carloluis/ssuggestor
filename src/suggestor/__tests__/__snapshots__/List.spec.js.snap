// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> snapshot - no suggestion list 1`] = `null`;

exports[`<List /> snapshot - suggestion list closed 1`] = `null`;

exports[`<List /> snapshot - suggestion list open (index 1) 1`] = `
<ul
  className="dropdown-menu ss-list"
>
  <li
    className=""
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      locellus
    </a>
  </li>
  <li
    className="ss-over-item"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      Lydian
    </a>
  </li>
  <li
    className=""
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      warray
    </a>
  </li>
</ul>
`;

exports[`<List /> snapshot - suggestion list open 1`] = `
<ul
  className="dropdown-menu ss-list"
>
  <li
    className="ss-over-item"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      locellus
    </a>
  </li>
  <li
    className=""
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      Lydian
    </a>
  </li>
  <li
    className=""
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <a>
      warray
    </a>
  </li>
</ul>
`;

exports[`<List /> snapshot 1`] = `null`;
